drv/lpc55-swd/src/main.rs

```diff
@@ -1058,21 +1297,626 @@ impl ServerImpl {
         }
 
         // Reset the STM32, causing it to reboot into the newly-set slot
-        use drv_lpc55_gpio_api::{Pins, Value};
-        let gpio = Pins::from(self.gpio);
-        gpio.set_val(ROT_TO_SP_RESET_L, Value::Zero);
+        self.sp_reset(false);
+        Ok(())
+    }
+
+    fn sp_reset(&mut self, squelch_notify: bool) {
+        self.sp_reset_enter();
         hl::sleep_for(10);
-        gpio.set_val(ROT_TO_SP_RESET_L, Value::One);
+        self.sp_reset_leave(squelch_notify);
+    }
+
+    fn sp_reset_enter(&mut self) {
+        setup_rot_to_sp_reset_l_out(self.gpio);
+        let gpio = Pins::from(self.gpio);
+        gpio.set_val(ROT_TO_SP_RESET_L_OUT, Value::Zero);
+    }
+
+    /// Cleanup is true if we don't want to be notified of this reset
+    fn sp_reset_leave(&mut self, squelch_notify: bool) {
+        let gpio = Pins::from(self.gpio);
+        setup_rot_to_sp_reset_l_in(self.gpio);
+        gpio.set_val(ROT_TO_SP_RESET_L_IN, Value::One); // should be a no-op
+        if squelch_notify {
+            let _ = gpio.pint_op(
+                ROT_TO_SP_RESET_L_IN_PINT_SLOT,
+                PintOp::Clear,
+                PintCondition::Rising,
+            );
+            let _ = gpio.pint_op(
+                ROT_TO_SP_RESET_L_IN_PINT_SLOT,
+                PintOp::Clear,
+                PintCondition::Falling,
+            );
+        }
+    }
+
+    // The SP is halted at its reset vector.
+    fn sp_measure_fast(&mut self) -> Result<[u8; 256 / 8], ()> {
+        // write program entry address to PC (R15)
+        // write top-of-stack address to MSP (R13)
+        // write start of image/vector table address(0x20000000) to VTOR (at 0xe000ed08)
+        // write the program to RAM
+        // write DHCSR to RUN (MATIC + C_DEBUGEN)
+        // poll for S_HALT or timeout
+
+        // Set SP's Program Counter
+        let sp_reset_vector =
+            u32::from_le_bytes(ENDOSCOPE_BYTES[4..=7].try_into().unwrap_lite());
+        if self
+            .do_write_core_register(Reg::Dr.into(), sp_reset_vector | 1)
+            .is_err()
+        {
+            ringbuf_entry!(Trace::WrotePcRegisterFail);
+            return Err(());
+        }
+
+        // Set SP's Stack Pointer
+        let sp_initial_sp =
+            u32::from_le_bytes(ENDOSCOPE_BYTES[0..=3].try_into().unwrap_lite());
+        if self
+            .do_write_core_register(Reg::Sp.into(), sp_initial_sp)
+            .is_err()
+        {
+            ringbuf_entry!(Trace::WroteSpRegisterFail);
+            return Err(());
+        }
```

Comment:
Can you add a citation or short note about where the `0..=3` and `4..=7` come from for those who don't have vector tables memorized? Looks like B1.5.2 and B1.5.3 in the ARMv7m manual? 

With `sp_reset_vector | 1`, isn't the `| 1` redundant? I think the compiler should be generating proper code to set the thumb bit.

Reply:
done